### ISSUE TITLE
First frame is drawn as full screen - onRenderedFirstFrame not reliable

### ISSUE THREAD

**Issue by natario1** [google/ExoPlayer#5222]
### Tldr
The first video frame should be drawn when the video dimensions are applied to the surface, **or at least** `onRenderedFirstFrame` should be reliable in hiding this wrong sized frame.

### Issue description
The first video frame is distortedly drawn over the whole match_parent surface, before `AspectRatioFrameLayout.setAspectRatio()` is called (so way before it is remeasured, surface recreated, ...). You can see this by putting an all threads breakpoint in setAspectRatio method.

In the project demo app, this is not visible if we use an `exo_shutter`. But
1. if you don't have one, this first frame flickers before ARFL and surface are remeasured
2. even if you have one, _sometimes_ it's not enough: when the `exo_shutter` is hidden, the surface is still drawing in full size because the ARFL still did not go through measure.

I could not reproduce (2) in the demo app, but it is clearly visible in another project. This flicker is terrible when entering new UIs. I am attaching a snapshot of the issue occurring in the demo app, with media Misc -> Dizzy (MP4). That is a wide video but for this frame, it is rendered as full screen, distorted.

<img width="576" alt="flicker" src="https://user-images.githubusercontent.com/15526561/49690091-8185df80-fb2b-11e8-8f67-f5e9ee89d801.png">

There seems to be some race condition due to which, sometimes, the very first `AspectRatioFrameLayout.onMeasure()` valid call (with aspect ratio set) happens after the `onRenderedFirstFrame` call, so the shutter is hidden before it should.

This makes `onRenderedFirstFrame` useless, it should be invoked after the surface has the correct dimensions, since it is basically an UI callback. In my own project, even `post`ing and waiting an UI cycle from there is not enough (surface is still fullscreen).

### Reproduction steps
- Use https://github.com/natario1/ExoPlayer/tree/flicker which sets the shutter to transparent
- Download Misc - Dizzy (_not sure this is needed, please do_)
- Open Misc - Dizzy
- Observe that, although it blends with the animation, the first frame is full screen. It is better seen after recording a screen video.

### Link to test content
This branch reproduces the issue https://github.com/natario1/ExoPlayer/tree/flicker. 
It sets shutter to transparent. However, as I said, in my project this happens even with an opaque
shutter, which means that `onRenderedFirstFrame` is not reliable for this.

### Version of ExoPlayer being used
head / v 2.9.2

### Device(s) and version(s) of Android being used
Nexus 5X, Android 8.1.0

**Comment by ojw28** [comment]
Thanks for the detailed description.

> if you don't have one, this first frame flickers before ARFL and surface are remeasured

I think we should not consider this case. The solution is to have a shutter. The threading model of the underlying platform makes it difficult to avoid needing one.

> even if you have one, sometimes it's not enough: when the exo_shutter is hidden, the surface is still drawing in full size because the ARFL still did not go through measure.

The ordering when using `PlayerView` is:

1. `onVideoSizeChanged` is called. `AspectRatioFrameLayout.requestLayout` is called, which calls `requestLayout`, which should post a layout pass.
1. `onRenderedFirstFrame` is called, which calls `setVisibilitiy(INVISIBLE)` on the shutter.

So for what you say to be true, the layout pass triggered by (1) must happen after the effects of the `setVisibility` call triggered in step (2). I'm somewhat surprised this can happen (I think I'd expect the underlying platform to post both, in which case the effects of the two calls would occur in order). I wonder if the layout in your particular project requires multiple layout passes, which might explain why you can reproduce the problem there but not in the demo app.

Can you provide us with a way to reproduce the problem with the shutter? Maybe a patch to the demo app that changes the layout to something similar to what's in your own project would be sufficient to reproduce the issue?

**Comment by natario1** [comment]
Maybe the fact that `ARFL` is remeasured, does not mean that the underlying Surface is synchronously resized ?

It is very hard for me to provide a patch to demo app, this flicker happens when entering a fragment with enter transitions, shared element transition (which the player is not a part of). The fragment hierarchy is Coordinator -> FrameLayout -> PlayerView, plus siblings, with no special coordinator magic.

Doesn't look like a layout issue to me, `ARFL.onMeasure()` is called when it should, just once (no multiple layout passes). It's just that the surface keeps being fullscreen for a while.

I have tried to debug the UI frames using a Choreographer

```kotlin
Choreographer.getInstance().postFrameCallback(object: Choreographer.FrameCallback {
            override fun doFrame(frameTimeNanos: Long) {
                frameNum++
                Log.e("ExoPlayerBug", "Frame number: $frameNum")
                Choreographer.getInstance().postFrameCallback(this)
            }
        })
```

And set breakpoints at `onVideoSizeChanged`, `onRenderedFirstFrame` and `onMeasure`, and everything seems to be like you said. The typical order is `onVideoSizeChanged` and `onRenderedFirstFrame` on frame X, then `onMeasure` (X+1).

Sometimes it's even better: `onVideoSizeChanged`, then `onMeasure` on X+1, then `onRenderedFirstFrame` on X+1 as well. I don't know if the presence of the debugger is hiding the issue.

**Comment by ojw28** [comment]
It's somewhat tricky for us to do anything if we can't reproduce the problem (in an app we have source code access to; and preferably a simple one). It's also a little unclear whether this is an ExoPlayer issue, as opposed to an issue in the underlying view system that only occurs for your specific use case.

So I think we probably still need some more information for this issue to be actionable.

**Comment by natario1** [comment]
Thank you, I will try to go deeper into this and get back to you.

**Comment by natario1** [comment]
The delay shows up when I use logcat instead of breakpoints.

```
2018-12-11 12:46:08.163 W/ExoPlayerBug: Frame number: 2111
2018-12-11 12:46:08.177 E/ExoPlayerBug: PlayerView onVideoSizeChanged. Informing the frame.
2018-12-11 12:46:08.177 E/ExoPlayerBug: PlayerView onRenderedFirstFrame. Hiding shutter.
2018-12-11 12:46:08.179 W/ExoPlayerBug: Frame number: 2112
2018-12-11 12:46:08.196 W/ExoPlayerBug: Frame number: 2113
2018-12-11 12:46:08.213 W/ExoPlayerBug: Frame number: 2114
2018-12-11 12:46:08.229 W/ExoPlayerBug: Frame number: 2115
2018-12-11 12:46:08.246 W/ExoPlayerBug: Frame number: 2116
2018-12-11 12:46:08.249 E/ExoPlayerBug: AspectRatioFrameLayout onMeasure. Video: 1.8181819
```

This points the finger at transitions so I have tried to change their duration. Turns out that `AspectRatioFrameLayout.onMeasure()` will never be called until there is a shared element transition running. It is executed only on the frame when the transition ends. Does anyone know what is the mechanism behind this?

The shared element transition targets only an `ImageView` which is not contained in PlayerView.

```xml
<CoordinatorLayout>
    <FrameLayout>
        <PlayerView/>
        <ImageView/> <!-- Shared element -->
    </FrameLayout>
</CoordinatorLayout>
``` 
I have tried changing `requestLayout` to `forceLayout` but it does nothing.

**Comment by natario1** [comment]
Well, I'm not able to fix this, will just call `playerView.setPlayer` when the transition is over.

The only thing the library could do better is in `PlayerView` docs where it could be made clearer that `exo_shutter` is a key component to avoid flickers & not just an opaque view that covers the video surface.

Feel free to close this, thanks for answering.

**Comment by ojw28** [comment]
> Does anyone know what is the mechanism behind this?

Afraid not. We're not particularly expert on the Android view framework. You might get a helpful answer if you ask on StackOverflow though!

I'll add an extra sentence to the `PlayerView` Javadoc where `exo_shutter` is documented, to explain that it's also helpful for preventing flicker when using `SurfaceView`. On that note, you might like to try using `TextureView` instead. `TextureView` also has `setTransform`, which may well be of use to you given you're doing clipping (https://github.com/google/ExoPlayer/issues/5221).
